### ISSUE TITLE
Binary data type

### ISSUE THREAD

**Issue by ghost** [dhall-lang/dhall-lang#418]
Has anyone implemented a binary data type?
I'm using dhall to implement a binary language, and since dhall has a binary specification, it would make sense to have binary types.

Side question: dhallToCbor in the cards?

**Comment by Gabriel439** [comment]
@lifranc: Dhall does not yet have support for binary values (which I will call `Bytes`)

The main reason why is because it doesn't come up that often in most configuration file formats and when it does come up you typically need a `Text`-like representation anyway so that it can be displayed and edited (such as base-64-encoded bytes or something similar)

**Comment by lifranc** [comment]
Yes, for now text strings work. I guess my question is 'Do you plan to implement an actual byte/hexadec/base32/base<insertInt> type eventually?'
I personally like base32 since it lines up with bytes, is fairly efficient and unambiguous. But anything would do. Thanks!

**Comment by Gabriel439** [comment]
@lifranc: Do you have a syntax in mind for `Bytes` literals?

**Comment by lifranc** [comment]
@Gabriel439 I'v never come across a byte literal. I guess I'd use [0..7] ?
Not very space efficient thou.
BTW, I just saw the new version of dhall already encodes/decodes to binary. Nice!

**Comment by Gabriel439** [comment]
@lifranc: Yeah, `dhall` already supports a standard conversion to CBOR expressions and binary, and you can find the full details here: https://github.com/dhall-lang/dhall-lang/blob/master/standard/binary.md

I think for now I'd probably still recommend requiring users to encode bytes in `Text` in some way

**Comment by singpolyma** [comment]
I would suggest maybe not a `Bytes` type, but a `Byte` type with syntax `0xFF` then `List Byte` can be written:

`[ 0x00, 0x01, 0xFF ]`

This is useful for two reasons: it allows the common hex syntax used for binary data (this is the main reason to use hex literal in other languages, is when treating numbers as bits or bytes) and it allows a type with a capped range (`List Natural` can encode the above, but also obviously invalid cases)

`dhall-to-yaml` could convert `List Byte` to the standard yaml `!binary` type

**Comment by f-f** [comment]
@singpolyma would this mean forbidding bound names like `0xFF`?

**Comment by Gabriel439** [comment]
I want to make sure we understand the use case for this.  The reason why is that YAML doesn't support bytes and it doesn't come up often in ops-related contexts.  In fact, most other configuration formats don't support bytes either.

**Comment by Nadrieril** [comment]
Even the YAML `!binary` type [actually works with a base64 string](https://yaml.org/type/binary.html). In a human-writable format like Dhall, I feel like a simple base64 string is actually much more convenient than a special Byte type would be since base64 can easily be converted to and from binary with common tools.

**Comment by lifranc** [comment]
@Gabriel439 I'll give the same advice as before; don't wait for the standard, but create it. That said you might still think binary types aren't so hot. This [presentation][1] shows a way out of our current mess into a sensible approach to computing. Ritchie and Thompson either cut the corner or didnt realize the enormous mistake of storing their new lines-of-text-UI to coding *as-is* instead of using a de/compiler that would store and retrieve it as *code*. Somehow this gigantic mistake was never fixed. Not yet anyhow!

[1]: https://www.youtube.com/watch?v=PBo1lEZ_Iik

**Comment by dumblob** [comment]
I actually totally agree with @lifranc - few years ago I even created configuration/serialization/streaming/you_name_it format ([uSX](https://github.com/dumblob/uSX )) whose purpose was to equally well support human readable version as well as binary data **without** reencoding them (i.e. fully zero-copy without any preparation/precompilation/whatever steps) and among other things support [streamed data](https://github.com/dhall-lang/dhall-lang/issues/594 ). uSX is to any other existing serialization/configuration format as [Rebol](http://www.rebol.com/ ) or [Red](https://www.red-lang.org/ ) or [Spry](http://sprylang.se/ ) or even [Lisp](https://en.wikipedia.org/wiki/Lisp_(programming_language) ) to programming languages - i.e. a micro meta framework with free form usage providing "endless" possibilities while guaranteeing the minimum required functionality and nothing more.

To wrap up, first class support for binary data (at best fully arbitrary data with zero-copy serialization and deserialization) is a must for a modern format.

**Comment by Gabriel439** [comment]
I'm going to request that people don't use GitHub reactions to comment on difficult discussions.  Doing so can shut down a productive conversation

@dumblob: So far we are currently focused on ops-related use cases and we want to make sure that those use cases are well-served before we branch out to address a broader audience.  This is why we currently spend most of our complexity budget on safety-related features, rather than performance.

**Comment by dumblob** [comment]
>So far we are currently focused on ops-related use cases and we want to make sure that those use cases are well-served before we branch out to address a broader audience. This is why we currently spend most of our complexity budget on safety-related features, rather than performance.

Understood. Focus on ops has undoubtedly precedence over performance. Let me though point out, that support for binary data is first of all not about performance, but rather about the discomfort of dealing with non-transparent conversion made by the user before putting it in dhall (typically an up front conversion using base64 or alike) - i.e. the need to additionally describe "hey, wait, I'm lying, this is not a string, it's actually a JPEG, but wait, actually not exactly JPEG, but temporarily base64 which you'll need to deal with yourself" in a side channel. In other words, this forces one to use dhall as pure meta language instead of a direct specification language.

**Comment by f-f** [comment]
@dumblob the clients on the other side of the wire will have to do something with the binary data (e.g. open it as a JPEG), so you'll have to agree on that something, and agreeing on a decoding pipeline like "decode Dhall → open as JPEG" doesn't sound very different to me than agreeing on "decode Dhall → decode base64 → open as JPEG" (if we don't have to worry about performance)

That is, you'll have to anyways agree over a side channel what the clients are supposed to do with the data. However, you can send some of this information in channel, e.g. by defining a union of different encodings that the client is supposed to decode:
```dhall
let Encoding = < Base64 | JPEG | PDF > -- etc etc. 

-- Note: The client will have to know about this from a side channel 
-- (e.g. you host it somewhere and both server and client import it with http)
```
..and then sending a list of operations together with the data:
```dhall
in
  { encoded = "some base64 encoded stuff here"
  , encodings = [ Encoding.Base64, Encoding.JPEG ] 
  }
```

**Comment by lifranc** [comment]
My main use case was to embed raw binary code, ie without translating it into dhall. Think next gen ops code which are always binary serialized. So mainly this is a practical type. I believe we are on a 32 bit standard and will probably remain so for at least our lifetime. And base 32 is unambiguous. Base 64  is also fine. Obviously UTF8 based languages arent really concerned about data size but about *readability* , so I like 32 bits here again.

**Comment by dumblob** [comment]
@f-f yep, I fully understand that as that's how it's being done in most of other existing formats. What differs is nothing more than our view on the significance of the difference between "decode Dhall → open as JPEG" and "decode Dhall → decode base64 → open as JPEG" as the latter one is way more error prone to implement, to follow and not easy to maintain (having e.g. 10 such different places is already an overkill :cry:).

But let's see how this will evolve in dhall.

**Comment by winitzki** [comment]
The `Bytes` type and import `as Bytes` are now fully implemented.
